304 - Ids Slider add component

vivz753 opened the pull request:
**Explain the details for making this change. What existing problem does the pull request solve?**
<!--
Example: When "Adding a function to do X",
explain why it is necessary to have a way to do X.
-->
Blocker: wanting for changes to eslintrc for ternaries and short circuits

**Related github/jira issue (required)**:
<!--
Provide a link to the related issue(s) to this Pull Request;
auto-closing github issues if necessary (example: "Closes #100" or "Fixes #1230")
-->
- Closes #304 

**Steps necessary to review your pull request (required)**:
<!--
Include:
- commands you ran and their output
- screenshots / videos
- test scenarios
-->

**Included in this Pull Request**:
- [ ] An e2e or functional test for the bug or feature.
- [ ] A note to the change log.

<!-- After submitting your PR, please check back to make sure checks on the PR -->

tmcconechy commented on src/components/ids-slider/README.md:
```diff
@@ -0,0 +1,128 @@
+
+## Terminology
+
+- Minimum: The smallest value possible, on one end of the slider
```
```suggestion
- **Minimum**: The smallest value possible, on one end of the slider
```

For all these please bold like the other examples

vivz753 replied:
Done

tmcconechy commented on src/components/ids-slider/README.md:
```diff
@@ -0,0 +1,128 @@
+- `step-number` {number >= 2} number of steps (applicable only for the step slider)
+- `vertical` {boolean}
+
+## Settings (Non-attributes)
```
Whats the different with attributes and none attributes ?

vivz753 replied:
this if for this.labels

it's technically not an attribute because we don't want arrays being passed through HTML as strings... so they would have to set it using javascript, hence "non-attribute"

tmcconechy replied:
ok thats what i guessed. Thanks

tmcconechy commented on src/components/ids-slider/README.md:
````diff
@@ -0,0 +1,128 @@
+In order for the labels to be set, the `type` attribute must be set to `step`, and the attribute `step-number` must match the size of the array of labels. Javascript is required to inject the label array into the DOM.
+
+```js
+this.labels = ['very bad, poor, average, good, excellent']; // where 'this' is the ids-slider
````
So there is no way to set the label in the DOM? Or is that just for the step ones because its an object/array

vivz753 replied:
Yeah this is for the step slider... currently the default labels are the min and max value

tmcconechy commented on src/components/ids-slider/README.md:
```diff
@@ -0,0 +1,128 @@
+
+## Description
+
+Holds the value(s) between a minimum and maximum, which can be controlled by the thumb(s) along the track of the slider. There is a single slider (1 thumb), double slider (2 thumbs), and a step slider (1 thumb and x amount of ticks). They can also be converted from horizontal (default) to vertical orientation. The value(s) can be modified by either clicking the area around the track, by dragging  ...
```
```suggestion
Displays a range of selectable values(s) between a minimum and maximum, which can be controlled by the thumb(s) along the track of the slider. There is a single slider (1 thumb), double slider (2 thumbs), and a step slider (1 thumb and x amount of ticks). They can also be converted from horizontal (default) to vertical orientation. The value(s) can be modified by either clicking the area around the track, by dragging the thumb back and forth, or by focusing and pressing the arrow keys.  A user can select a single value or range of values. 
```

tmcconechy commented on src/components/ids-slider/README.md:
```diff
@@ -0,0 +1,128 @@
+
+## Use Cases
+
+- Control an input with limited options through clicking/dragging/keyboard interaction
```
```suggestion
- Control an input with limited options through clicking/dragging/keyboard interaction.
- . Best for showing a user’s current selection relative to the lower and upper limits of a scale.
```

vivz753 replied:
Done

tmcconechy commented on src/components/ids-slider/ids-slider.d.ts:
```diff
@@ -0,0 +1,27 @@
+import { IdsElement } from '../../core';
+
+export default class IdsSlider extends IdsElement {
+  /** Sets the color theme of the chart */
```
```suggestion
  /** Sets the color of the slider bar */
```

tmcconechy commented on src/components/ids-slider/ids-slider.js:
```diff
@@ -0,0 +1,1052 @@
+
+  #percentSecondary;
+
+  slider;
```
These are all fine. But what i have been doing is not declaring them and just doing this to declare them. Not sure if this is better but a minor point to consider.

```
this.slider = this.container.querySelector('.slider');
```

vivz753 replied:
That's what I was doing initially, but Robert suggested initializing them in the class just for better readability to know what variables can be used

tmcconechy replied:
ok i guess not a bad idea but then they maybe should have comments.

tmcconechy commented on src/components/ids-slider/ids-slider.js:
```diff
@@ -0,0 +1,1052 @@
+    }
+  }
+
+  get isRtl() {
```
```suggestion
  get isRTL() {
```

vivz753 replied:
Done

tmcconechy commented on src/components/ids-slider/ids-slider.js:
```diff
@@ -0,0 +1,1052 @@
+    this.onEvent('languagechange.container', this.closest('ids-container'), async (e) => {
+      await this.setLanguage(e.detail.language.name);
+      const isRtl = this.locale.isRTL(e.detail.language.name);
+      this.isRtl = isRtl;
```
Maybe make it all RTL like the locale api. Is there any need to have this.isRTL when there is this.locale.isRTL?

vivz753 replied:
I've changed `isRtl` to `isRTL` ... I suppose I could technically perform all the updates in the event listener for language change, but it's nice to keep it encapsulated and refer to isRTL in my own class.. I'm not exactly sure what you mean, but I'm open to suggestions if you can provide examples

tmcconechy replied:
I mean if you need to check for isRTL anywhere in the code you could use 

```
if (this.locale.isRTL())
```

So `this.isRTL` may be redundant?

vivz753 replied:
I have events that I need to trigger when RTL gets changed so in my set isRTL() I recalculate the boundaries of the component, update the position of thumbs and progress bar.. I think it's simpler for me to keep that encapsulated there... 

I think I can get rid of the `await this.setLanguage(e.detail.language.name)` though

tmcconechy replied:
That implementation is part of the local mixin. Once its mixed on you get the locale functions.
You don't have to pass the language in. If you don't it will take the correct language.

tmcconechy commented on src/components/ids-slider/ids-slider.js:
```diff
@@ -0,0 +1,1052 @@
+    this.type === 'double' && this.#addDragEvents('secondary');
+
+    // KEYBOARD EVENTS
+    this.#addKeyboardEvents();
```
```suggestion
    this.#addKeyboarListeners();
```

Did this on the other PR... for a name

vivz753 replied:
I'll commit it on my end since it's not spelled correctly here

vivz753 replied:
Should this be `attachKeyboardListeners()` instead of `add`?

tmcconechy replied:
Yes my bad as per https://github.com/infor-design/enterprise-wc/pull/321#issue-723790999 

attachEventHandlers and attachKeyboardListeners

tmcconechy commented on src/components/ids-slider/ids-slider.js:
```diff
@@ -0,0 +1,1052 @@
+    // console.log('this.trackArea.offsetTop: ' + this.trackArea.offsetTop) // 0
+    // console.log('this.trackArea.offsetLeft: ' + this.trackArea.offsetLeft) // -12
+    // console.log('this.trackArea.clientWidth: ' + this.trackArea.clientWidth) // 24
+    // console.log('this.trackArea.clientHeight: ' + this.trackArea.clientHeight) // 24
```
Remove these comments?

vivz753 replied:
> @vivz753 the PR with the new lint rules is in so you can merge it on. Did a surface pass on this as well see comments.
> 
> 1. Check inline comments
> 2. On the example im only seeing the vertical slider. Do you have examples for the others to add?
> 3. When clicking the slider I see a blue area which wasnt in the design (but i dont hate it)
> 
> <img alt="Screen Shot 2021-09-03 at 9 42 25 AM" width="107" src="https://user-images.githubusercontent.com/814283/132015008-1c20a6cd-6edd-4eb3-9ec1-231f8e403829.png">
> 
> 4. When i tab into it and use up/down arrow the keys appear to be reversed (up is down/ down is up)

1. What do you mean check inline comments? 
2. Updating the examples with the other sliders now
3. The blue area hover was just to help me develop, I can remove it
4. The reason why up/down is like that is because I have plans to reverse the min/max to be bottom/top (right now it's min/max top/bottom)

tmcconechy replied:
1. I mean look at the comments i put on the code inline (you did)
2. cool
3. sure or can make it a setting
4. ok cool as long as its the right way in the end... (up goes up and down goes down / left decreased / right increases)

vivz753 replied:
Finished making changes, pretty much all ready to go--it took a bit longer than expected bc I discovered some bugs where calculating the x/y screwed up when scrolling away from the initial window position.

tmcconechy reviewed (commented):
@vivz753 the PR with the new lint rules is in so you can merge it on. Did a surface pass on this as well see comments.

1. Check inline comments
2. On the example im only seeing the vertical slider. Do you have examples for the others to add?
3. When clicking the slider I see a blue area which wasnt in the design (but i dont hate it)
<img width="107" alt="Screen Shot 2021-09-03 at 9 42 25 AM" src="https://user-images.githubusercontent.com/814283/132015008-1c20a6cd-6edd-4eb3-9ec1-231f8e403829.png">
4. When i tab into it and use up/down arrow the keys appear to be reversed (up is down/ down is up)

vivz753 commented on demos/ids-slider/example.js:
```diff
@@ -0,0 +1,3 @@
+document.addEventListener('DOMContentLoaded', () => {
+  document.querySelector('.survey').labels = ['very bad', 'poor', 'average', 'good', 'excellent'];
```
I need help getting this to work.. this file isn't linking up to my `example.html` for some reason

How does this example.js file link to the html in the other component demos? @tmcconechy

tmcconechy replied:
@vivz753 I think the example.js is linked. Its firing...for me. Maybe restart the demo app?

```
document.addEventListener('DOMContentLoaded', () => {
  alert('DOMContentLoaded');
  document.querySelector('.survey').labels = ['very bad', 'poor', 'average', 'good', 'excellent'];
});
alert('in the js');
```

EdwardCoyle replied:
@vivz753 I was able to set a breakpoint on this line and see it trigger.  Could just need to restart 👍

vivz753 replied:
Thanks, that worked

tmcconechy commented:
@vivz753 a couple more things since now i see the examples more

1. If i do left and right arrow on the horizontal splitters it scrolls the whole page
2. the numbers on step slider doesn't quite line up
3. should also be able to click the numbers like https://main-enterprise.demo.design.infor.com/components/slider/example-index.html
4. should make the "pointer" cursor show on the bar/ticks and slider

That being said I am impressed with the work on this PR.

jmacaluso711 commented on src/components/ids-slider/ids-slider.scss:
```diff
@@ -0,0 +1,212 @@
+        bottom: 0%; // vert
+      }
+
+      background-color: #0072ED;
```
Should probably use our token here `--ids-color-palette-azure-90`

vivz753 replied:
this is actually `--ids-color-palette-asure-60` and `--ids-color-brand-primary-base`

going with the `ids-color-brand-primary-base`

jmacaluso711 commented on src/components/ids-slider/ids-slider.scss:
```diff
@@ -0,0 +1,212 @@
+.ids-slider {
+  &[mode='contrast'] {
+    .tick {
+      background-color: #004A99;
```
Same as above as well as on lines 205 and 209

vivz753 replied:
Done

EdwardCoyle reviewed (commented):
@vivz753 This looks great!  Willing to approve after addressing Tim's comments, along with this:

- On the 4.x Range (double) Slider, the tooltip is only supposed to appear on the "active" thumb.  I don't really dislike seeing both values when manipulating the slider, but wondering if we should either put it behind a setting or limit the tooltip to just the "active" thumb for the sake of matching the current functionality.

jmacaluso711 commented on src/components/ids-slider/ids-slider.scss:
```diff
@@ -0,0 +1,212 @@
+
```
One general comment for this file: Nested styles should have 1 line space above and 1 below (if not the last declaration in the block). Not sure why the linter didn't catch it. For example lines 9-10

jmacaluso711 reviewed (changes requested):
@vivz753 I left a few minor comments related to your .scss file. Also looks like there are a few js lint errors you need to address. A bunch are related to ternary and line length, but there are some that are unrelated, like spacing issues.

vivz753 commented:
> @vivz753 a couple more things since now i see the examples more
> 
>     1. If i do left and right arrow on the horizontal splitters it scrolls the whole page
> 
>     2. the numbers on step slider doesn't quite line up
> 
>     3. should also be able to click the numbers like https://main-enterprise.demo.design.infor.com/components/slider/example-index.html
> 
>     4. should make the "pointer" cursor show on the bar/ticks and slider
> 
> 
> That being said I am impressed with the work on this PR.

Thank you for the compliment!!

1. added e.preventDefault() to stop scrolling when pressing arrow keys
2. added justify-content:center to center labels for non-RTL and RTL
3. added functionality to click labels to update value
4. added cursor pointer css on track area and labels

vivz753 commented:
> @vivz753 This looks great! Willing to approve after addressing Tim's comments, along with this:
> 
>     * On the 4.x Range (double) Slider, the tooltip is only supposed to appear on the "active" thumb.  I don't really dislike seeing both values when manipulating the slider, but wondering if we should either put it behind a setting or limit the tooltip to just the "active" thumb for the sake of matching the current functionality.

Yeah I noticed that but still intentionally added the visibility for both. It made more sense imo to show both--why should a user have to constantly click back and forth just to see/check both values? That's 2 actions and then them having to remember the previous value to make sure they entered precisely the correct numbers. Doesn't seem like a good design to me... but I acknowledge that may have been on purpose for whatever reason.

Thank you for reviewing @EdwardCoyle

tmcconechy commented on demos/ids-slider/example.html:
```diff
@@ -0,0 +1,39 @@
+    <ids-slider type="single"></ids-slider>
+  </ids-layout-grid-cell>
+  <ids-layout-grid-cell>
+    <ids-text font-size="12" type="h1">Double Slider: </ids-text>
```
I think we should call "double" "range" so lets refer to the two handle slider as a range slider as before https://main-enterprise.demo.design.infor.com/components/slider/example-range.html

tmcconechy commented on src/components/ids-slider/TODO.md:
```diff
@@ -0,0 +1,69 @@
+    - [x] make sure double range slider sets default inputs even if attribute valueb is unspecified or out of the bounds of min/max
+    - [x] test double range slider with negative min and positive max
+    - [x] write tests
+  - [-] bugs
```
It might make sense too just have the points that are left open so there isnt so much in here as a TODO thats already done

tmcconechy commented on src/components/ids-slider/ids-slider.scss:
```diff
@@ -0,0 +1,220 @@
+      position: absolute;
+      width: 0; 
+      height: 0; 
+      bottom: -6px; // horizontal
```
Maybe all these //horiz comments can be formed into more complete comments

tmcconechy commented on src/components/ids-slider/ids-slider.js:
```diff
@@ -0,0 +1,1044 @@
+   * @param {string} value The color, this can be a hex code with the #, a native css color, or an ids-status color
+   */
+  set color(value) {
+    this.setAttribute(attributes.COLOR, value);
```
I noticed this means you can pass in "green" but we want to stick to (or at last support our color pallete variables) i tihnk do the color setter like ids-text https://github.com/infor-design/enterprise-wc/blob/main/src/components/ids-text/ids-text.js#L216-L234

Then update the example to `ruby-60` `emerald-60` ect

deep7102 reviewed (approved):
Looks very nice, Thanks! @vivz753
